runtime/lua/vim/filetype/detect.lua

```diff
@@ -821,7 +954,50 @@ function M.sys(bufnr)
   end
 end
 
-function M.tex(path, bufnr) end
+-- Choose context, plaintex, or tex (LaTeX) based on these rules:
+-- 1. Check the first line of the file for "%&<format>".
+-- 2. Check the first 1000 non-comment lines for LaTeX or ConTeXt keywords.
+-- 3. Default to "plain" or to g:tex_flavor, can be set in user's vimrc.
+function M.tex(path, bufnr)
+  local format = getlines(bufnr, 1):find('^&%s*(%a+)')
+  if format then
+    format = format:lower():gsub('pdf', '', 1)
+    if format == 'tex' then
+      vim.bo[bufnr].filetype = 'tex'
+    elseif format == 'plaintex' then
+      vim.bo[bufnr].filetype = 'plaintex'
+    end
+  elseif path:lower():find('tex/context/.*/.*%.tex') then
+    vim.bo[bufnr].filetype = 'context'
+  else
+    local lpat = [[documentclass\>\|usepackage\>\|begin{\|newcommand\>\|renewcommand\>]]
+    local cpat =
+      [[start\a\+\|setup\a\+\|usemodule\|enablemode\|enableregime\|setvariables\|useencoding\|usesymbols\|stelle\a\+\|verwende\a\+\|stel\a\+\|gebruik\a\+\|usa\a\+\|imposta\a\+\|regle\a\+\|utilisemodule\>]]
+
+    for _, line in ipairs(getlines(bufnr, 1, 1000)) do
```

Comment:
This doesn't seem to match what filetype.vim is doing. filetype.vim

1. finds the first non-comment line in the first 1000 lines, then
2. from there, search through the next 1000 lines for a `lpat` or `cpat` match.

This implementation is just searching the first 1000 lines regardless. It's probably an edge case, but we should still try and catch it.